src/sentry/middleware/integrations/parsers/msteams.py

```diff
@@ -0,0 +1,56 @@
+from __future__ import annotations
+
+import logging
+
+from django.http import HttpResponse
+from django.urls import resolve
+
+from sentry.middleware.integrations.parsers.base import BaseRequestParser
+from sentry.models.integrations.integration import Integration
+from sentry.models.outbox import WebhookProviderIdentifier
+from sentry.services.hybrid_cloud.util import control_silo_function
+from sentry.types.integrations import EXTERNAL_PROVIDERS, ExternalProviders
+
+logger = logging.getLogger(__name__)
+
+
+class MsTeamsRequestParser(BaseRequestParser):
+    provider = EXTERNAL_PROVIDERS[ExternalProviders.MSTEAMS]
+    webhook_identifier = WebhookProviderIdentifier.MSTEAMS
+
+    @control_silo_function
+    def get_integration_from_request(self) -> Integration | None:
+        request = self.request
+        data = request.data  # type:ignore
+        integration = None
+        try:
+            payload = data["value"]["payload"]
+            integration_id = payload["integrationId"]
+            integration = Integration.objects.filter(id=integration_id).first()
+        except Exception:
+            pass
+
+        if integration:
+            return integration
+
+        try:
+            channel_data = data["channelData"]
+            team_id = channel_data["team"]["id"]
+            integration = Integration.objects.filter(
+                provider=self.provider, external_id=team_id
+            ).first()
+        except Exception:
+            pass
+        return integration
+
+    def get_response(self) -> HttpResponse:
+        result = resolve(self.request.path)
+        if result.url_name != "sentry-integration-msteams-webhooks":
+            return self.get_response_from_control_silo()
+
+        regions = self.get_regions_from_organizations()
+        if len(regions) == 0:
+            logger.error("no_regions", extra={"path": self.request.path})
+            return self.get_response_from_control_silo()
```

Comment:
So the getting the regions from the orgs, which we get from the integration is the only way we send data to the region silos. That means the two patterns for getting an `Integration` from the request (`data.value.payload.integrationId` and `data.channelData.team.id`) are the only methods for forwarding to regions. 

Do all the payloads intended for regions follow this pattern? Because if not, it'll attempt to handle the webhook at the control silo, and may fail.

Reply:
@leeandher I went through all the code paths for the webhooks, and recycled the payloads in `tests/sentry/integrations/msteams/test_webhook.py` to create some test cases for what we expect to be run in the region and control silos.

There are some webhooks for which we do not need to fetch the `Integration` object for. In these instances, we should safely be able to process them at the control silo.